src/xmp_error.rs

```diff
@@ -0,0 +1,269 @@
+// Copyright 2022 Adobe. All rights reserved.
+// This file is licensed to you under the Apache License,
+// Version 2.0 (http://www.apache.org/licenses/LICENSE-2.0)
+// or the MIT license (http://opensource.org/licenses/MIT),
+// at your option.
+
+// Unless required by applicable law or agreed to in writing,
+// this software is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR REPRESENTATIONS OF ANY KIND, either express or
+// implied. See the LICENSE-MIT and LICENSE-APACHE files for the
+// specific language governing permissions and limitations under
+// each license.
+
+use std::{ffi::CStr, fmt};
+
+use num_enum::FromPrimitive;
+use thiserror::Error;
+
+use crate::ffi::CXmpError;
+
+/// Describes error conditions returned by most XMP Toolkit operations.
+#[derive(Debug)]
+#[non_exhaustive]
+pub struct XmpError {
+    /// A selector for the specific error type.
+    pub error_type: XmpErrorType,
+
+    /// Descriptive string, for debugging use only. It must not be shown to users in a final
+    /// product. It is written for developers, not users, and never localized.
+    pub debug_message: String,
+}
+
+impl XmpError {
+    pub(crate) fn raise_from_c(err: &CXmpError) -> XmpResult<()> {
+        if err.had_error != 0 {
+            Err(XmpError {
+                error_type: XmpErrorType::from(err.id),
+                debug_message: if err.debug_message.is_null() {
+                    String::default()
+                } else {
+                    unsafe {
+                        CStr::from_ptr(err.debug_message)
+                            .to_string_lossy()
+                            .into_owned()
+                    }
+                },
+            })
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl fmt::Display for XmpError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        if self.debug_message.is_empty() {
+            write!(f, "XmpError({})", self.error_type)
+        } else {
+            write!(f, "XmpError({}, {})", self.error_type, self.debug_message)
+        }
+    }
+}
+
+impl std::error::Error for XmpError {}
+
+/// Describes which error type occurred.
+#[derive(Debug, Error, FromPrimitive, PartialEq)]
+#[non_exhaustive]
+#[repr(i32)]
+pub enum XmpErrorType {
+    // --- Generic error codes ---
+    /// Generic unknown error.
+    #[error("Unknown error")]
+    #[num_enum(default)]
+    Unknown = 0,
```

Comment:
It the error is not in this enum set could we use Unknown(u32) to return the actual error value.  Otherwise I am not sure how many other error would just show up as Unknown.

Reply:
Ironically, there is a specific unknown error with the selector 0 as defined here:

https://github.com/adobe/XMP-Toolkit-SDK/blob/main/public/include/XMP_Const.h#L1371

But … the point is taken, I could create another enum variant that would cover other cases. My thinking was that any such case reflects something happening in the C++ toolkit that was not predicted, so `Unknown` was actually a good place to put it.

Reply:
Leaving this as is and merging.